cli/azd/pkg/tools/bicep/bicep.go

```diff
@@ -5,170 +5,225 @@ package bicep
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"io"
 	"log"
+	"net/http"
 	"os"
-	"os/user"
 	"path/filepath"
 	"runtime"
-	"strings"
 
-	"github.com/azure/azure-dev/cli/azd/pkg/convert"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/azure/azure-dev/cli/azd/pkg/config"
 	"github.com/azure/azure-dev/cli/azd/pkg/exec"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
 	"github.com/azure/azure-dev/cli/azd/pkg/tools"
 	"github.com/blang/semver/v4"
 )
 
+// cBicepVersion is the minimum version of bicep that we require (and the one we fetch when we fetch bicep on behalf of a
+// user).
+var cBicepVersion semver.Version = semver.MustParse("0.12.40")
+
 type BicepCli interface {
-	tools.ExternalTool
 	Build(ctx context.Context, file string) (string, error)
 }
 
-func NewBicepCli(commandRunner exec.CommandRunner) BicepCli {
-	return &bicepCli{
-		commandRunner: commandRunner,
-	}
+// NewBicepCli creates a new BicepCli. Azd manages its own copy of the bicep CLI, stored in `$AZD_CONFIG_DIR/bin`. If
+// bicep is not present at this location, or if it is present but is older than the minimum supported version, it is
+// downloaded.
+func NewBicepCli(
+	ctx context.Context,
+	console input.Console,
+	commandRunner exec.CommandRunner,
+) (BicepCli, error) {
+	return newBicepCliWithTransporter(ctx, console, commandRunner, http.DefaultClient)
 }
 
-type bicepCli struct {
-	commandPath   *string
-	commandRunner exec.CommandRunner
-}
+// newBicepCliWithTransporter is like NewBicepCli but allows providing a custom transport to use when downloading the
+// bicep CLI, for testing purposes.
+func newBicepCliWithTransporter(
+	ctx context.Context,
+	console input.Console,
+	commandRunner exec.CommandRunner,
+	transporter policy.Transporter,
+) (BicepCli, error) {
+	if override := os.Getenv("AZD_BICEP_TOOL_PATH"); override != "" {
+		log.Printf("using external bicep tool: %s", override)
+
+		return &bicepCli{
+			path:   override,
+			runner: commandRunner,
+		}, nil
+	}
 
-func (cli *bicepCli) Name() string {
-	return "Bicep CLI"
-}
+	bicepPath, err := cachedBicepPath()
+	if err != nil {
+		return nil, fmt.Errorf("finding bicep: %w", err)
+	}
 
-func (cli *bicepCli) InstallUrl() string {
-	return "https://aka.ms/azure-dev/bicep-install"
-}
+	if _, err = os.Stat(bicepPath); err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, fmt.Errorf("finding bicep: %w", err)
+	} else if errors.Is(err, os.ErrNotExist) {
```

Comment:
since the previous `if` is returning, we can avoid the `else if` and just have the next `if` below

Reply:
Thanks! I fixed this up.